brainframe_qt/ui/resources/ui_elements/applications/messaging_application.py

```diff
@@ -30,25 +30,40 @@ def __init__(self, *, socket_name: str):
 
         self._init_signals()
 
-    def _init_message_server(self, socket_name: str) \
+    def _init_message_server(self, socket_name: str, retries: int = 0) \
             -> Optional["MessagingServer"]:
+
+        # Check if there's already an active server that is ready to communicate
+        if MessagingSocket.is_server_alive(socket_name):
+            return None
+
         message_server = MessagingServer(socket_name=socket_name, parent=self)
 
+        # Note that this will only trigger on Linux:
+        # https://doc.qt.io/qt-5/qlocalserver.html#listen
+        # """
+        #   Note: On Unix if the server crashes without closing listen will fail with
+        #   AddressInUseError. To create a new server the file should be removed. On
+        #   Windows two local servers can listen to the same pipe at the same time, but
+        #   any connections will go to one of the server.
+        # """
         if message_server.serverError() == QAbstractSocket.AddressInUseError:
-            if not MessagingSocket.is_server_alive(socket_name):
-                logging.warning(f"Socket for IPC is open, but connection was "
-                                "refused. Attempting takeover.")
-                # Attempt hostile takeover of socket
-                MessagingServer.removeServer(socket_name)
-                message_server = self._init_message_server(socket_name)
-
-                if message_server:
-                    logging.info("IPC server socket takeover successful.")
-                else:
-                    logging.error("Unable to perform takeover on socket.")
-            else:
+            if retries > 3:
+                # Already retried. Give up trying to create the server.
+                logging.error(f"")
                 return None
 
+            logging.warning(f"Socket for IPC is open, but connection was refused. "
+                            f"Attempting takeover.")
+            # Attempt hostile takeover of socket
+            MessagingServer.removeServer(socket_name)
+            message_server = self._init_message_server(socket_name, retries=retries + 1)
+
+            if message_server:
+                logging.info("IPC server socket takeover successful.")
+            else:
+                logging.error("Unable to perform takeover on socket.")
```

Comment:
Should this make the client exit?

Reply:
Yeah I considered that, but I figured that could be a bad idea. I have no idea what scenario would take us down this codepath, and I figured that the potential for a client that mysteriously never opens is worse than a client that's not a singleton

Reply:
Good point